### PR TITLE
[TINKERPOP-3110] Fix Bytecode build logic for strategies

### PR DESCRIPTION
**Pull request by saikiranboga**
JIRA: https://issues.apache.org/jira/browse/TINKERPOP-3110

### CODE REVIEW

File: CHANGELOG.asciidoc [hunk 1]
```diff
@@ -38,6 +38,7 @@ This release also includes changes from <<release-3-6-8, 3.6.8>>.
 * Added getter method to `ConcatStep`, `ConjoinStep`, `SplitGlobalStep` and `SplitLocalStep` for their private fields.
 * Gremlin Server docker containers shutdown gracefully when receiving a SIGTERM.
 * TINKERPOP-3080 Support to specify Operator as a reducer in withSideEffect when parsing with the grammar
+* TINKERPOP-3110 Fixed bug in Bytecode build logic for strategies
```
Comment (spmallette) [inline CHANGELOG.asciidoc:41]: the JIRA reference should not be in the CHANGELOG (i know its in the entry above, but that's not correct either). JIRA references are added on release in a separate listing.

**Comment by saikiranboga** [comment]
> i can't help wondering if you need additional tests. is there an existing test that ensures that we don't have duplicate strategies in bytecode anymore?

@spmallette, Added few more tests. We do have some tests around strategy validation, but not for duplicate checking.

**Comment by spmallette** [comment]
thanks for the fixes - VOTE +1

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/BytecodeHelper.java [hunk 2]
```diff
@@ -397,6 +418,8 @@ else if (arguments[i] instanceof Bytecode)
         }
     }
 
+
```
Comment (kenhuuu) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/BytecodeHelper.java:421]: very small nit: remove added spaces

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/BytecodeHelper.java [hunk 3]
```diff
@@ -350,6 +351,26 @@ public static <A extends TraversalStrategy> Iterator<A> findStrategies(final Byt
                 os -> (A) os.getArguments()[0]);
     }
 
+    public static <A extends TraversalStrategy> boolean removeStrategies(final Bytecode bytecode, final String operator, final Class<TraversalStrategy>[] clazzes) {
```
Comment (kenhuuu) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/BytecodeHelper.java:354]: Is `<A extends TraversalStrategy>`  needed? `A` doesn't seem to be used.

**Comment by kenhuuu** [comment]
Is there a test for the problem stated in TINKERPOP-3110? A test that asserts that if multiple with() are used that it is combined into a single OptionsStrategy that contains all the options from all the with()s?

**Comment by xiazcy** [comment]
Thanks @saikiranboga for the changes! Note you may need to resolve some CHANGELOG conflicts. 

VOTE +1

**Comment by saikiranboga** [comment]
> Is there a test for the problem stated in TINKERPOP-3110? A test that asserts that if multiple with() are used that it is combined into a single OptionsStrategy that contains all the options from all the with()s?

Yes. See `BytecodeHelperTest#shouldFindStrategy`.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/BytecodeHelper.java (hunk 3 above)
Comment (saikiranboga) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/BytecodeHelper.java]: Thanks for noticing. Removed.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/BytecodeHelper.java (hunk 2 above)
Comment (saikiranboga) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/util/BytecodeHelper.java]: Removed.

**Comment by kenhuuu** [comment]
VOTE +1

**Pull request merged**